### ISSUE TITLE
Type errors involving function types should show a better error message

### ISSUE THREAD

**Issue by itssamuelrowe** [pallene-lang/pallene#231]
The compiler generates "expected function but found function in assignment" for the following source code:
```
function f(): float
    return 3.14
end

function test(g: () -> integer)
    g = f
end
```
Basically, the compiler is complaining that `f` does not have a type that is compatible with `string -> ()`, which is the shape of `io_write`. Compatibility means that the function types have to be identical except that any is compatible with any other type.

The routine to print function names in `types.lua` has a TODO to print the full function type instead of just "function".

**Comment by hugomg** [comment]
That test case with `io_write` has more to do with #232. I suggest this one instead:

```
function f(): float
    return 3.14
end

function test(g: () -> integer)
    g = f
end
```

**Comment by hugomg** [comment]
Fixed in #239
